### PR TITLE
Markdown WYSIWYG support for posts and comments!

### PR DESCRIPTION
**Pull request by th-bunratta**
According to the original planned feature in the fourth iteration of the proposal, this will introduce more user-friendly site UI to the app's users.

Additions are as follows:

- The Markdown editor widget (w/ SimpleMDE)
  - Users can now write the threads in Markdown and see live preview of their Markdown source on:
     - Posts
     - Comments (including replies)

Also, this includes additional bug fixes to the previous mergers and, hopefully, will not introduce any new bugs as follows:
  - Posts that are broken during rendering (as in https://github.com/KooCook/koocook-dj/commit/f8cf2ec45f7ca764d86ea877dae23ea3576e1eb7.)
  - Posts that are not properly escaped, resulting in the disappearance of the body. (as in https://github.com/KooCook/koocook-dj/commit/d229e7868e0877b44a4673af54f0fc9e47376b50.)

Hope this helps enhance the app!

**PR Prerequisite to:** (in sequence)
 1. `feature/rating`
 2. 
     - `feature/popularity`
     - `feature/personalisation`

### CODE REVIEW

File: koocook_core/models/recipe.py [hunk 1]
```diff
@@ -53,5 +53,6 @@ def nutrition(self):
 
     @property
     def recipe_ingredients(self):
+
         """ Proxy property for consistency with Schema.org's standard """
```
Comment (MaiMee1) [inline koocook_core/models/recipe.py]: There's some style problem / typo here?

File: koocook_core/tests/test_support/test_markdown.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+# TODO: Preliminary tests
+class TestMarkdown(unittest.TestCase):
+    def test_proper_escape(self):
+        source = MarkdownSource("\"**Test**\"")
+        self.assertEqual(source.rendered, "<p>&quot;<strong>Test</strong>&quot;</p>")
+
+    def test_escaped(self):
+        source = MarkdownSource("&lt;&gt;")
+        self.assertEqual(source.rendered, "<p>&lt;&gt;</p>")
```
Comment (MaiMee1) [inline koocook_core/tests/test_support/test_markdown.py]: More tests would be welcomed. Is this PR a requirement for others?

Comment (th-bunratta) [reply koocook_core/tests/test_support/test_markdown.py]: PR Prerequisite to: (in sequence)

- `feature/rating`
- `feature/popularity`
- `feature/personalisation`

File: koocook_core/models/recipe.py (hunk 1 above)
Comment (th-bunratta) [reply koocook_core/models/recipe.py]: Perhaps, yes? I'll fix it up.

File: koocook_core/tests/test_support/test_markdown.py (hunk 2 above)
Comment (MaiMee1) [reply koocook_core/tests/test_support/test_markdown.py]: 😢

Comment (th-bunratta) [reply koocook_core/tests/test_support/test_markdown.py]: I've added unit tests! Feel free to check them out.

File: koocook_core/tests/test_support/test_markdown.py [hunk 3]
```diff
@@ -4,12 +4,86 @@
...
+    def setUp(self) -> None:
+        self.text = "This is a no markdown text"
+
+    def test_no_markdown_characters(self):
+        no_md_text = "*"
+        source = MarkdownSource(no_md_text)
+        self.assertEqual(source.rendered, f"<p>{no_md_text}</p>")
+        no_md_text = "`"
+        source = MarkdownSource(no_md_text)
+        self.assertEqual(source.rendered, f"<p>{no_md_text}</p>")
```
Comment (MaiMee1) [inline koocook_core/tests/test_support/test_markdown.py]: Testing this way, once one test fails, the other doesn't even run. I suggest using `subTest`s

File: koocook_core/tests/test_support/test_markdown.py [hunk 4]
```diff
@@ -4,12 +4,86 @@
...
 
 
-# TODO: Preliminary tests
-class TestMarkdown(unittest.TestCase):
-    def test_proper_escape(self):
+def load_article():
+    import json, os
+    dir_path = os.path.dirname(os.path.realpath(__file__))
+    article = json.loads(open(os.path.join(dir_path, "test_markdown/article.json")).read())
+    return article
```
Comment (MaiMee1) [inline koocook_core/tests/test_support/test_markdown.py]: No reason to import in a module method?

File: koocook_core/tests/test_support/test_markdown.py [hunk 5]
```diff
@@ -4,12 +4,86 @@
...
+        self.assertEqual(source.rendered, f"<p>{no_md_text}</p>")
+
+    def test_markdown_bold_tag(self):
+        text = "Never really over"
+        md_text = f"**{text}**"
+        source = MarkdownSource(md_text)
+        self.assertEqual(source.rendered, f"<p><strong>{text}</strong></p>")
+        md_text = f"__{text}__"
+        source = MarkdownSource(md_text)
+        self.assertEqual(source.rendered, f"<p><strong>{text}</strong></p>")
```
Comment (MaiMee1) [inline koocook_core/tests/test_support/test_markdown.py]: Same thing here as above

File: koocook_core/tests/test_support/test_markdown.py [hunk 6]
```diff
@@ -4,12 +4,86 @@
...
+
+    def test_markdown_italic_tag(self):
+        text = "Never Really Over"
+        md_text = f"*{text}*"
+        source = MarkdownSource(md_text)
+        self.assertEqual(source.rendered, f"<p><em>{text}</em></p>")
```
Comment (MaiMee1) [inline koocook_core/tests/test_support/test_markdown.py]: What about using `_` for italics?

File: koocook_core/tests/test_support/test_markdown.py [hunk 7]
```diff
@@ -4,12 +4,86 @@
...
+    def test_markdown_with_html(self):
+        no_md_text = f"<strong>{self.text}</strong> **{self.text}**"
+        source = MarkdownSource(no_md_text)
+        self.assertEqual(source.source, f"__{self.text}__ **{self.text}**")
+        self.assertEqual(source.rendered, f"<p><strong>{self.text}</strong> <strong>{self.text}</strong></p>")
+        no_md_text = f"*{self.text}* <em>{self.text}</em> **{self.text}**"
+        source = MarkdownSource(no_md_text)
+        self.assertEqual(source.source, f"*{self.text}* _{self.text}_ **{self.text}**")
+        self.assertEqual(source.rendered, f"<p><em>{self.text}</em> <em>{self.text}</em> "
+                                          f"<strong>{self.text}</strong></p>")
```
Comment (MaiMee1) [inline koocook_core/tests/test_support/test_markdown.py]: I don't think `no_md_text` is a correct descriptive name. (In fact I believe it's misleading?) (It says "no md" but there is md! You see stars and such in there.)

File: koocook_core/tests/test_support/test_markdown.py [hunk 8]
```diff
@@ -4,12 +4,86 @@
...
+
+    # This prevents malicious attacks
+    def test_html_escape(self):
         source = MarkdownSource("\"**Test**\"")
```
Comment (MaiMee1) [inline koocook_core/tests/test_support/test_markdown.py]: Using outer single quotes is better than using escapes.

File: koocook_core/tests/test_support/test_markdown.py [hunk 9]
```diff
@@ -4,12 +4,86 @@
...
+        source = MarkdownSource("<script></script>")
+        self.assertEqual(source.rendered, "<p>&lt;script&gt;&lt;/script&gt;</p>")
 
-    def test_escaped(self):
+    def test_html_escaped(self):
         source = MarkdownSource("&lt;&gt;")
         self.assertEqual(source.rendered, "<p>&lt;&gt;</p>")
+
+        source = MarkdownSource("&quot; *&lt;&gt;* __&amp;__")
+        self.assertEqual(source.rendered, "<p>&quot; <em>&lt;&gt;</em> <strong>&amp;</strong></p>")
```
Comment (MaiMee1) [inline koocook_core/tests/test_support/test_markdown.py]: Again, for multiple cases using `subTest` would be better

**Comment by th-bunratta** [comment]
> I think data should be separated where it can be grouped? Like a data directory for tests that has subdirectories and what not. Putting test data in the test folder itself without labelling it as data seems messy. The name of the directory you used also overlaps with the test file (`.../test_support/test_markdown.py` and `.../test_support/test_markdown/article.json`). While this shouldn't cause problem currently, I don't think it is a great idea.

It's there waiting to integrate with your fixtures in another branch.

**Pull request merged**